File: test/clj/tales/api_test.clj
(ns tales.api_test
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [clojure.test :refer :all]
            [me.raynes.fs :as fs]
            [ring.mock.request :as mock]
            [ring.util.io :refer [string-input-stream]]
            [ring.util.response :refer [get-header]]
            [tales.api :as api]
            [tales.handler :refer [app]]
            [tales.test-utility :refer [content-type?
                                        content-type-json?
                                        temporary-projects]]))

(use-fixtures :each temporary-projects)

(deftest test-api-get-projects
  (testing "returns empty list for empty projects"
    (let [response (app (mock/request :get "/api/tales"))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 200 (:status response)))
      (is (content-type-json? response))
      (is (empty? body))))

  (testing "returns existing projects"
    (let [project1 (:body (api/create-project {:name "Project 1"}))
          project2 (:body (api/create-project {:name "Project 2"}))
          project3 (:body (api/create-project {:name "Project 3"}))
          response (app (mock/request :get "/api/tales"))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 200 (:status response)))
      (is (content-type-json? response))
      (is (= 3 (count body)))
      (is (some #{project1} body))
      (is (some #{project2} body))
      (is (some #{project3} body)))))

(deftest test-api-get-project
  (testing "returns not-found for non-existing project"
    (let [response (app (mock/request :get "/api/tales/unknown"))]
      (is (= 404 (:status response)))
      (is (content-type-json? response))))

  (testing "returns project"
    (let [project (:body (api/create-project {:name "Test"}))
          response (app (mock/request :get "/api/tales/test"))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 200 (:status response)))
      (is (content-type-json? response))
      (is (= project body)))))

(deftest test-api-create-project
  (testing "fails for invalid params"
    (let [response (app (-> (mock/request :post "/api/tales")
                          (mock/json-body {})))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 400 (:status response)))
      (is (content-type-json? response))
      (is (not (nil? (:error body))))))

  (testing "creates resource"
    (let [response (app (-> (mock/request :post "/api/tales")
                          (mock/json-body {:name "Test"})))
          location (get-header response "Location")]
      (is (= 201 (:status response)))
      (is (content-type-json? response))
      (is (clojure.string/ends-with? location "/api/tales/test"))))

  (testing "only creates resource once"
    (let [_ (api/create-project {:name "Existing"})
          response (app (-> (mock/request :post "/api/tales")
                          (mock/json-body {:name "Existing"})))
          location (get-header response "Location")]
      (is (= 409 (:status response)))
      (is (content-type-json? response))
      (is (not location))))

  (testing "returns the created project"
    (let [response (app (-> (mock/request :post "/api/tales")
                          (mock/json-body {:name "My Tale"})))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 201 (:status response)))
      (is (content-type-json? response))
      (is (= "my-tale" (:slug body)))
      (is (= "My Tale" (:name body))))))

(deftest test-api-update-project
  (testing "returns not-found for non-existing project"
    (let [response (app (-> (mock/request :put "/api/tales/unknown")
                          (mock/json-body {:name "My Tale"})))]
      (is (= 404 (:status response)))
      (is (content-type-json? response))))

  (testing "fails for invalid params"
    (let [_ (api/create-project {:name "Test"})
          response (app (-> (mock/request :put "/api/tales/test")
                          (mock/json-body {})))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 400 (:status response)))
      (is (content-type-json? response))
      (is (not (nil? (:error body))))))

  (testing "returns the updated project"
    (let [_ (api/create-project {:name "Test"})
          response (app (-> (mock/request :put "/api/tales/test")
                          (mock/json-body {:name "My Tale"})))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 200 (:status response)))
      (is (content-type-json? response))
      (is (= "test" (:slug body)))
      (is (= "My Tale" (:name body))))))

(deftest test-api-delete-project
  (testing "returns not-found for non-existing project"
    (let [response (app (mock/request :delete "/api/tales/unknown"))]
      (is (= 404 (:status response)))
      (is (content-type-json? response))))

  (testing "deletes resource"
    (let [_ (api/create-project {:name "Test"})
          first-try (app (mock/request :delete "/api/tales/test"))
          second-try (app (mock/request :delete "/api/tales/test"))]
      (is (= 200 (:status first-try)))
      (is (content-type-json? first-try))
      (is (= 404 (:status second-try)))
      (is (content-type-json? second-try)))))

(deftest test-image-type
  (testing "recognizes gif images"
    (is (= "gif" (api/image-type "image/gif"))))
  (testing "recognizes gif images"
    (is (= "png" (api/image-type "image/png"))))
  (testing "recognizes png images"
    (is (= "jpg" (api/image-type "image/jpeg"))))
  (testing "recognizes bmp images"
    (is (= "bmp" (api/image-type "image/bmp"))))
  (testing "recognizes svg images"
    (is (= "svg" (api/image-type "image/svg+xml"))))
  (testing "fails for unknown image types"
    (is (not (api/image-type nil)))
    (is (not (api/image-type "image/foo")))))

(def example-file (str/join "\n" ["<svg version=\"1.1\""
                                  "  baseProfile=\"full\""
                                  "  width=\"300\" height=\"200\""
                                  "  xmlns=\"http://www.w3.org/2000/svg\">"
                                  "    <circle cx=\"150\" cy=\"100\" r=\"80\""
                                  "      fill=\"green\" />"
                                  "</svg>"]))

(deftest test-api-update-project-image
  (testing "returns not-found for non-existing project"
    (let [response (app (-> (mock/request :put "/api/tales/unknown/image")
                          (mock/content-type "image/svg+xml")
                          (mock/content-length (count example-file))
                          (mock/body example-file)))]
      (is (= 404 (:status response)))
      (is (content-type-json? response))))

  (testing "returns bad-request for unsupported content-type"
    (let [_ (api/create-project {:name "Test"})
          response (app (-> (mock/request :put "/api/tales/test/image")
                          (mock/content-type "text/plain")
                          (mock/content-length (count example-file))
                          (mock/body example-file)))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 400 (:status response)))
      (is (content-type-json? response))
      (is (= "Invalid content-type: text/plain" (:error body)))))

  (testing "copies uploaded file to project directory"
    (let [_ (api/create-project {:name "Test"})
          response (app (-> (mock/request :put "/api/tales/test/image")
                          (mock/content-type "image/svg+xml")
                          (mock/content-length (count example-file))
                          (mock/body example-file)))
          target-file (fs/file tales.project/*project-dir* "test" "test.svg")]
      (is (= 200 (:status response)))
      (is (content-type-json? response))
      (is (fs/exists? target-file))))

  (testing "returns the updated project"
    (let [_ (api/create-project {:name "Test"})
          response (app (-> (mock/request :put "/api/tales/test/image")
                          (mock/content-type "image/svg+xml")
                          (mock/content-length (count example-file))
                          (mock/body example-file)))
          body (json/read-str (:body response) :key-fn keyword)]
      (is (= 200 (:status response)))
      (is (content-type-json? response))
      (is (= "test" (:slug body)))
      (is (= "test.svg" (:file-path body))))))

File: pkg/web/api_test.go
package web

import (
	"encoding/json"
	"io"
	"net/http"
	"testing"

	"github.com/stretchr/testify/assert"

	"synyx.de/tales/pkg/project"
)

func TestAPI_listProjects(t *testing.T) {
	t.Run("get empty list of projects", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("GET", "/api/tales/", nil)

		assert.Equal(t, 200, resp.StatusCode)
		assert.Equal(t, "application/json; charset=utf-8", resp.Header.Get("Content-Type"))

		var projects []project.Project
		body, _ := io.ReadAll(resp.Body)
		assert.Nil(t, json.Unmarshal(body, &projects))
		assert.Empty(t, projects)
	})
	t.Run("get list of projects", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		_, _ = tc.repo.SaveProject("test-1", project.Project{})
		_, _ = tc.repo.SaveProject("test-2", project.Project{})
		_, _ = tc.repo.SaveProject("test-3", project.Project{})

		resp := tc.Request("GET", "/api/tales/", nil)

		assert.Equal(t, 200, resp.StatusCode)
		assert.Equal(t, "application/json; charset=utf-8", resp.Header.Get("Content-Type"))

		var projects []project.Project
		body, _ := io.ReadAll(resp.Body)
		assert.Nil(t, json.Unmarshal(body, &projects))
		assert.Len(t, projects, 3)
	})
	t.Run("internal error", func(t *testing.T) {
		tc := NewTestClient(t)
		tc.Cleanup()

		resp := tc.Request("GET", "/api/tales/", nil)

		AssertHTTPError(t, resp, 500)
	})
}

func TestAPI_createProject(t *testing.T) {
	t.Run("create project by slug", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		p := project.Project{Slug: "foo", Name: "Bar"}
		resp := tc.Request("POST", "/api/tales/", p)

		AssertProjectResponse(t, resp, p, 201)
	})
	t.Run("create project by name", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		p := project.Project{Name: "Foo Bar"}
		resp := tc.Request("POST", "/api/tales/", p)

		AssertProjectResponse(t, resp, project.Project{Slug: "foo-bar", Name: "Foo Bar"}, 201)
	})
	t.Run("conflicting project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		_, _ = tc.repo.SaveProject("foo", project.Project{Slug: "foo", Name: "Bar"})

		p := project.Project{Slug: "foo", Name: "Bar"}
		resp := tc.Request("POST", "/api/tales/", p)

		AssertError(t, resp, "project already exists\n", 409)
	})
	t.Run("invalid project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("POST", "/api/tales/", nil)

		AssertError(t, resp, "invalid project\n", 400)
	})
	t.Run("unnamed project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("POST", "/api/tales/", project.Project{})

		AssertError(t, resp, "invalid project\n", 400)
	})
	t.Run("internal error", func(t *testing.T) {
		tc := NewTestClient(t)
		tc.Cleanup()

		resp := tc.Request("POST", "/api/tales/", project.Project{Slug: "foo"})

		AssertError(t, resp, tc.dir+" not accessible", 500)
	})
}

func TestAPI_loadProject(t *testing.T) {
	t.Run("existing project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		p := project.Project{Slug: "foo"}
		_, _ = tc.repo.SaveProject("foo", p)

		resp := tc.Request("GET", "/api/tales/foo", nil)

		AssertProjectResponse(t, resp, p, 200)
	})
	t.Run("unknown project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("GET", "/api/tales/foo", nil)

		AssertError(t, resp, "project not found\n", 404)
	})
	t.Run("internal error", func(t *testing.T) {
		tc := NewTestClient(t)
		tc.Cleanup()

		resp := tc.Request("GET", "/api/tales/foo", nil)

		AssertError(t, resp, tc.dir+" not accessible", 500)
	})
}

func TestAPI_updateProject(t *testing.T) {
	t.Run("update project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		_, _ = tc.repo.SaveProject("foo", project.Project{Slug: "foo", Name: "Bar"})

		p := project.Project{Slug: "foo", Name: "Baz"}
		resp := tc.Request("PUT", "/api/tales/foo", p)

		AssertProjectResponse(t, resp, p, 202)
	})
	t.Run("invalid project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("PUT", "/api/tales/foo", nil)

		AssertError(t, resp, "invalid project", 400)
	})
	t.Run("unknown project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		p := project.Project{Slug: "foo"}
		resp := tc.Request("PUT", "/api/tales/foo", p)

		AssertError(t, resp, "project not found\n", 404)
	})
	t.Run("internal error", func(t *testing.T) {
		tc := NewTestClient(t)
		tc.Cleanup()

		resp := tc.Request("PUT", "/api/tales/foo", project.Project{Slug: "foo"})

		AssertError(t, resp, tc.dir+" not accessible", 500)
	})
}

func TestAPI_deleteProject(t *testing.T) {
	t.Run("existing project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		_, _ = tc.repo.SaveProject("foo", project.Project{Slug: "foo"})

		resp := tc.Request("DELETE", "/api/tales/foo", nil)

		assert.Equal(t, 202, resp.StatusCode)

		body, _ := io.ReadAll(resp.Body)
		assert.Empty(t, string(body))
	})
	t.Run("unknown project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("DELETE", "/api/tales/foo", nil)

		AssertError(t, resp, "project not found", 404)
	})
	t.Run("internal error", func(t *testing.T) {
		tc := NewTestClient(t)
		tc.Cleanup()

		resp := tc.Request("DELETE", "/api/tales/foo", nil)

		AssertError(t, resp, tc.dir+" not accessible", 500)
	})
}

func TestAPI_updateProjectImage(t *testing.T) {
	t.Run("update project image", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		_, _ = tc.repo.SaveProject("foo", project.Project{Slug: "foo", Name: "Bar"})

		data := []byte{'f', 'o', 'o'}
		header := http.Header(map[string][]string{})
		header.Add("Content-Type", "image/bmp")
		resp := tc.RequestWithHeaders("PUT", "/api/tales/foo/image", data, header)

		AssertProjectResponse(t, resp, project.Project{Slug: "foo", Name: "Bar", FilePath: "foo.bmp",
			FileType: "image/bmp"}, 202)
	})
	t.Run("invalid content type", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		_, _ = tc.repo.SaveProject("foo", project.Project{Slug: "foo", Name: "Bar"})

		resp := tc.Request("PUT", "/api/tales/foo/image", nil)

		AssertError(t, resp, "unsupported content-type", 500)
	})
	t.Run("unknown project", func(t *testing.T) {
		tc := NewTestClient(t)
		defer tc.Cleanup()

		resp := tc.Request("PUT", "/api/tales/foo/image", nil)

		AssertError(t, resp, "project not found", 404)
	})
}

func AssertProjectResponse(t *testing.T, resp *http.Response, expected project.Project, statusCode int) {
	assert.Equal(t, statusCode, resp.StatusCode)
	assert.Equal(t, "application/json; charset=utf-8", resp.Header.Get("Content-Type"))

	var received project.Project
	body, _ := io.ReadAll(resp.Body)
	assert.NoError(t, json.Unmarshal(body, &received))
	assert.Equal(t, expected, received)
}

func AssertError(t *testing.T, resp *http.Response, errorMsg string, statusCode int) {
	AssertHTTPError(t, resp, statusCode)
	body, _ := io.ReadAll(resp.Body)
	assert.Contains(t, string(body), errorMsg)
}

func AssertHTTPError(t *testing.T, resp *http.Response, statusCode int) {
	assert.Equal(t, statusCode, resp.StatusCode)
	assert.Equal(t, "text/plain; charset=utf-8", resp.Header.Get("Content-Type"))
}
